crates/accelerate/src/target_transpiler/qargs.rs

```diff
@@ -0,0 +1,278 @@
+    type Item = Qargs;
+
+    type IntoIter = IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        IntoIter { inner: self }
+    }
+}
+
+impl<'py> IntoPyObject<'py> for TargetQargs {
+    type Target = PyAny;
+
+    type Output = Bound<'py, PyAny>;
+
+    type Error = PyErr;
+
+    fn into_pyobject(self, py: pyo3::Python<'py>) -> Result<Self::Output, Self::Error> {
+        match self {
+            Self::Global => Ok(py.None().into_bound(py)),
+            Self::Concrete(small_vec) => Ok(PyTuple::new(py, small_vec)?.into_any()),
+        }
+    }
+}
+
+impl<'py> FromPyObject<'py> for TargetQargs {
+    fn extract_bound(ob: &Bound<'py, PyAny>) -> PyResult<Self> {
+        let qargs: Option<Qargs> = ob.extract()?;
+        Ok(qargs.into())
+    }
+}
+
+impl Equivalent<TargetQargs> for TargetQargsRef<'_> {
+    fn equivalent(&self, key: &TargetQargs) -> bool {
+        *self == key.as_ref()
+    }
+}
+
+/// Reference representation of [TargetQargs].
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Default)]
+pub enum TargetQargsRef<'a> {
+    #[default]
+    Global,
+    Concrete(&'a [PhysicalQubit]),
+}
+
+impl<'a> TargetQargsRef<'a> {
+    /// Checks if the qargs in question are `Global`.
+    pub fn is_global(&self) -> bool {
+        matches!(self, Self::Global)
+    }
+
+    /// Checks if the qargs in question are `Concrete`.
+    pub fn is_concrete(&self) -> bool {
+        !self.is_global()
+    }
+
+    /// Returns an iterator of either zero or 1 step depending on whether
+    /// the operation is global (0) or concrete (1)
+    pub fn iter(self) -> Iter<'a> {
+        Iter { inner: self }
+    }
+
+    /// Turns the qargs into an option view
+    pub fn as_option(&self) -> Option<&'a [PhysicalQubit]> {
+        match self {
+            TargetQargsRef::Global => None,
+            TargetQargsRef::Concrete(qargs) => Some(qargs),
+        }
+    }
+
+    /// Returns the enclosed qargs in the case of `Concrete` variant.
+    ///
+    /// This function may `panic!`, and its unsafe use is discouraged.
+    pub fn unwrap(self) -> &'a [PhysicalQubit] {
+        match self {
+            Self::Global => panic!("Attempted to unwrap a 'Global' variant of 'TargetQargs'"),
+            Self::Concrete(qargs) => qargs,
+        }
+    }
+
+    /// Returns the enclosed qargs in the case of `Concrete` variant, otherwise
+    /// return an empty slice.
+    pub fn unwrap_or_default(self) -> &'a [PhysicalQubit] {
+        match self {
+            Self::Global => &[],
+            Self::Concrete(qargs) => qargs,
+        }
+    }
+
+    /// Converts a reference version of [TargetQargsRef] into an owned version.
+    pub fn to_owned(&self) -> TargetQargs {
+        match self {
+            TargetQargsRef::Global => TargetQargs::Global,
+            TargetQargsRef::Concrete(physical_qubits) => {
+                TargetQargs::Concrete(physical_qubits.to_owned().into())
+            }
+        }
+    }
```

Comment:
This isn't perfect (should be the trait impl), but I see the problem with `Borrow`.  I suspect this means that there's a better way to do the modelling of `Qargs`, but I'm not immediately clear what it is.

Reply:
Exploring...